src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/ScanOperationBase.cs

```diff
@@ -2,27 +2,29 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System.Linq;
-using System.Threading.Tasks;
-using Microsoft.DotNet.Darc.Options.VirtualMonoRepo;
+using Microsoft.DotNet.Darc.Options;
 using Microsoft.DotNet.DarcLib.VirtualMonoRepo;
 using Microsoft.Extensions.DependencyInjection;
+using System;
+using System.Linq;
+using System.Threading.Tasks;
 
-#nullable enable
 namespace Microsoft.DotNet.Darc.Operations.VirtualMonoRepo;
 
-internal class ScanOperation : Operation
+internal abstract class ScanOperationBase : Operation
 {
-    public ScanOperation(ScanCommandLineOptions options) : base(options, options.RegisterServices())
+    public ScanOperationBase(CommandLineOptions options, IServiceCollection services = null) : base(options, services)
     {
     }
 
     public override async Task<int> ExecuteAsync()
     {
-        var vmrScanner = Provider.GetRequiredService<IVmrScanner>();
+        var vmrCloakedFileScanner = Provider.GetServices<IVmrScanner>().First(scanner => scanner.GetType() == GetScannerType());
```

Comment:
In my opinion it's better to make `ScanOperationBase<T> : Operation where T : IVmrScanner` and then `Provider.GetService<T>()` instead of the `GetType()` dance. Similar pattern is used for the `VmrInitializer` and `VmrUpdater` though we relaxed the `where` there recently.

Reply:
I think I'd still have to do the `GetType()` dance or register the scanners without the interface (`service.TryAddTransient<VmrCloakedFileScanner>()`). In the case of `VmrInitializer` and `VmrUpdater` they're registered with their specific interfaces `IVmrInitializer` and `IVmrUpdater`, and they're the only implementations of that interface. If I my classes inherited the `ScanOperationBase<IVmrScanner>`, then they'd always be resolved to the last `IVmrScanner` registered

Reply:
You can make `IBinaryVmrScanner : IVmrScanner` and do `BinaryScanner : VmrScannerBase<IBinaryVmrScanner>`

It's always better to have the compiler type check for you than to pass some magic string like "binary" and hope it matches the class name (what if the class is renamed)

Reply:
You need to the register `<IBinaryVmrScanner, BinaryVmrScanner>` of course

Reply:
`service.TryAddTransient<VmrCloakedFileScanner>()` and do `ScanOperationBase<T> : Operation where T : VmrScanner`. That's still better than the string for type name

Reply:
Yup, I did that in the end. Doesn't feel right creating empty interfaces, might be wrong tho

Reply:
It's a thing when you need to mock the classes but we don't care about that in this case as we don't mock the top-level classes.